app/src/main/java/com/github/sdpteam15/polyevents/EditProfileActivity.kt

```diff
@@ -81,40 +81,59 @@ class EditProfileActivity : AppCompatActivity() {
             val newName = name.text.toString()
-            val newRank = UserRole.valueOf(rank.text.toString())
+            val newRank = stringToRank(rank.text.toString())
 
-            if(profile.userRole != newRank)
-                map[DatabaseConstant.PROFILE_RANK] = newRank.toString()
-            if(profile.profileName != newName)
-                map[DatabaseConstant.PROFILE_NAME] = newName
+            profile.userRole = newRank
+            profile.profileName = newName
 
-            currentDatabase.updateProfile(map, pid).observe {
-                if(it.value)
+            currentDatabase.updateProfile(profile).observe {
+                if (it.value) {
+                    end.postValue(true, this)
                     onBackPressed()
+                }
                 else
-                    HelperFunctions.showToast(getString(R.string.EditProfileActivity_DatabaseError),this)
+                    HelperFunctions.showToast(
+                        getString(R.string.EditProfileActivity_DatabaseError),
+                        this
+                    )
             }
         }
     }
 
-    fun rankOnFocusChangeListener(b :Boolean) {
+    fun rankOnFocusChangeListener(b: Boolean) {
         if (b) {
             rank.setText("")
         } else {
-            try {
-                lastRank = UserRole.valueOf(rank.text.toString())
-            } catch (e: Exception) {
-                rank.setText(lastRank.toString())
-            }
+            lastRank = stringToRank(rank.text.toString())
+            rank.setText(rankToString(lastRank))
         }
     }
 
-    fun nameOnFocusChangeListener(b :Boolean) {
+    fun nameOnFocusChangeListener(b: Boolean) {
         if (!b) {
-            if(name.text.toString().isNotEmpty())
+            if (name.text.toString().isNotEmpty())
                 lastName = name.text.toString()
             else
                 name.setText(lastName)
         }
     }
+
+    fun rankToString(rank: UserRole): String {
+        when (rank) {
+            UserRole.ADMIN -> return resources.getStringArray(R.array.EditProfileActivity_Ranks)[0]
+            UserRole.ORGANIZER -> return resources.getStringArray(R.array.EditProfileActivity_Ranks)[1]
+            UserRole.STAFF -> return resources.getStringArray(R.array.EditProfileActivity_Ranks)[2]
+            UserRole.PARTICIPANT -> return resources.getStringArray(R.array.EditProfileActivity_Ranks)[3]
+        }
+    }
+
+    fun stringToRank(rank: String): UserRole {
+        when (rank) {
+            resources.getStringArray(R.array.EditProfileActivity_Ranks)[0] -> return UserRole.ADMIN
+            resources.getStringArray(R.array.EditProfileActivity_Ranks)[1] -> return UserRole.ORGANIZER
+            resources.getStringArray(R.array.EditProfileActivity_Ranks)[2] -> return UserRole.STAFF
+            resources.getStringArray(R.array.EditProfileActivity_Ranks)[3] -> return UserRole.PARTICIPANT
+        }
+        return lastRank
+    }
```

Comment:
rankToString can be set as the "ToString" method of UserRole enum class
I'm not sure why you use StringToRank instead of the constructor of UserRole that takes a string too

Reply:
I will if you found how to get AppCompatActivity.ressources in UserRole

Reply:
Oh I see I haven't thought of that. You're right.